charts/metrics-server/templates/apiservice.yaml

```diff
@@ -1,17 +1,46 @@
-{{- if .Values.apiService.create -}}
+{{- $altNames := list }}
+{{- $certs := dict }}
+
+{{- if eq .Values.tls.type "helm" }}
+{{- $commonName := include "metrics-server.fullname" . }}
+{{- $ns := .Release.Namespace }}
+{{- $altNames = append $altNames (printf "%s.%s" $commonName $ns) }}
+{{- $altNames = append $altNames (printf "%s.%s.svc" $commonName $ns) }}
+{{- $altNames = append $altNames (printf "%s.%s.svc.%s" $commonName $ns .Values.tls.clusterDomain) }}
+{{- $certs = genSelfSignedCert $commonName nil $altNames 36500 }}
```

Comment:
Please look up the existing secret instead of generating a new certificate each time the chart is deployed.

Reply:
We can easily recreate the Secrets every helm upgrade. It's even more secure to change it on a regular basis IMHO.
Using the `lookup` function isn't safe against helm tooling as already mentioned somewhere in this PR.

People who are using
- Terraform template datasource
- Argo CD
- probably [Helmfile](https://helmfile.readthedocs.io/) _(cannot verify my statement here as I don't us this)_
- probably [Tanka](https://tanka.dev/)_(cannot verify my statement here as I don't us this)_

uses the helm chart only for templating. And template cannot use `lookup`

Reply:
I did not test the upgrade procedure. But it should be fine:
- We update both the Secret and the APIserver resource
- Kubernetes spin up a new Pod and performs a rolling update
- The existing Pod should have an updated Secret mounted as we don't use subPath mounting options. MS uses generic `k8s.io/apiserver/pkg/server` to implement the API, this one uses fsnotify to watch if certificates changes:
  - https://github.com/kubernetes/apiserver/blob/00a0da2cb88bcb3c971bc419510c50ca8b1b7abe/pkg/server/dynamiccertificates/dynamic_serving_content.go#L148
- ...

Do you need some test results here? :-)

Reply:
We probably need to test that MS is reloading the cert when it changes and then what the behaviour during the cert change is. If there is no interruption then I think we could not lookup the cert but add a comment in the code to say it could be done.

Reply:
This has been my experience with optimising Fluent Bit for hot-reload.

If the Helm pattern used `lookup` it could be documented as interruption safe which run directly from Helm, if interruptions are an issue and you're using an alternative apply method then use Cert Manager (or an external secret).

Reply:
Another option would be to use an init container or a helm job hook and something like ingress-nginx from the kubernetes org:
https://github.com/kubernetes/ingress-nginx/tree/main/images/kube-webhook-certgen
Seems to be a fork of https://github.com/jet/kube-webhook-certgen

Reply:
This seems like overengineering to me. For the Helm implementation we should use `lookup` and document why you might not want to do so.

Reply:
Alright. Lookup is already implemented 👍